src/catchup/CatchupManagerImpl.cpp

```diff
@@ -79,59 +88,78 @@ CatchupManagerImpl::processLedger(LedgerCloseData const& ledgerData)
-        {
-            // Case 1: At the ledger where we should start buffering.
-            addToSyncingLedgers(ledgerData);
-            firstLedgerInBuffer = lastReceivedLedgerSeq;
-        }
-        else
+        if (!mSmallestBufferedCheckpoint ||
+            *mSmallestBufferedCheckpoint > lastReceivedLedgerSeq)
         {
-            // Case 2: Still waiting to start buffering.
-            firstLedgerInBuffer =
-                hm.firstLedgerAfterCheckpointContaining(lastReceivedLedgerSeq);
+            mSmallestBufferedCheckpoint =
+                make_optional<uint32_t>(lastReceivedLedgerSeq);
         }
     }
-    else
-    {
-        // Case 3: Already buffering, keep buffering.
-        addToSyncingLedgers(ledgerData);
-        firstLedgerInBuffer = mSyncingLedgers.front().getLedgerSeq();
-    }
-    assert(hm.isFirstLedgerInCheckpoint(firstLedgerInBuffer));
 
-    // Finally we wait some number of ledgers beyond the first buffered
-    // ledger before we trigger the CatchupWork. This could be any number,
-    // for the sake of simplicity at the moment it's set to one ledger
-    // after the first buffered one.
+    // Finally we wait some number of ledgers beyond the smallest buffered
+    // checkpoint ledger before we trigger the CatchupWork. This could be any
+    // number, for the sake of simplicity at the moment it's set to one ledger
+    // after the first buffered one. Since we can receive out of order ledgers,
+    // we just check for any ledger larger than the checkpoint
+
     std::string message;
-    uint32_t catchupTriggerLedger =
-        hm.ledgerToTriggerCatchup(firstLedgerInBuffer);
-    if (lastReceivedLedgerSeq >= catchupTriggerLedger)
+
+    uint32_t lastLedgerInBuffer = mSyncingLedgers.crbegin()->first;
+    if (mSmallestBufferedCheckpoint &&
+        *mSmallestBufferedCheckpoint < lastLedgerInBuffer)
     {
         message = fmt::format("Starting catchup after ensuring checkpoint "
                               "ledger {} was closed on network",
-                              catchupTriggerLedger);
+                              lastLedgerInBuffer);
+
+        // We only need ledgers starting from the checkpoint. We can
+        // remove all ledgers before this
+        auto it = mSyncingLedgers.find(*mSmallestBufferedCheckpoint);
+        assert(it != mSyncingLedgers.end());
+        mSyncingLedgers.erase(mSyncingLedgers.begin(), it);
+
         startOnlineCatchup();
+
+        // Reset when we start catchup. If catchup fails, trimSyncingLedgers
+        // will set this again if a checkpoint still exists in mSyncingLedgers
+        mSmallestBufferedCheckpoint.reset();
     }
     else
     {
-        auto eta = (catchupTriggerLedger - lastReceivedLedgerSeq) *
+        uint32_t requiredFirstLedgerInCheckpoint =
+            mSmallestBufferedCheckpoint
+                ? *mSmallestBufferedCheckpoint
+                : hm.firstLedgerAfterCheckpointContaining(lastLedgerInBuffer);
```

Comment:
I don't understand this: why do we use the checkpoint derived from the last ledger in buffer, shouldn't it be the checkpoint derived from the first ledger instead as this is what we're going to use?

Reply:
if we have gaps where we're missing the checkpoint ledger, it's probably fine to give the ETA assuming we're going to recover from that gap (so aim to get the smallest checkpoint). I guess for this particular scenario, the ETA should be 0 as we should be getting that ledger "any time".

Reply:
I took the approach of always looking forward, but being optimistic and expecting the checkpoint ledger makes sense since we should be getting it.